### ISSUE TITLE
method reference (no macro) and custom data fails on type mismatch

### ISSUE THREAD

**Issue by lipanski** [nickel-org/nickel.rs#351]
hi!

I'm trying to use method references (as opposed to macros) **in order to split my app into multiple files/modules**. I'm using custom server data (the `Config` struct in this case).

this is a rough sketch of what I want to achieve:

```rust
#[macro_use]
extern crate nickel;

use nickel::{Nickel, HttpRouter, Request, Response, MiddlewareResult};

struct Config {}

fn hello_world<'mw>(_req: &mut Request, res: Response<'mw>) -> MiddlewareResult<'mw> {
    res.send("Hello World")
}

fn main() {
    let mut server = Nickel::with_data(Config {});
    server.get("**", hello_world);
    server.listen("127.0.0.1:6767");
}
```

however, the code fails with:

```
main.rs:14:12: 14:15 error: type mismatch: the type `fn(&mut nickel::Request<'_, '_>, nickel::Response<'mw>) -> std::result::Result<nickel::Action<nickel::Response<'mw>, nickel::Response<'mw, (), hyper::net::Streaming>>, nickel::NickelError<'mw>> {hello_world}` implements the trait `for<'r, 'r, 'r, 'mw> std::ops::Fn<(&'r mut nickel::Request<'r, 'r>, nickel::Response<'mw>)>`, but the trait `for<'r, 'mw, 'conn> std::ops::Fn<(&'r mut nickel::Request<'mw, 'conn, Config>, nickel::Response<'mw, Config>)>` is required (expected struct `Config`, found ()) [E0281]
main.rs:14     server.get("**", hello_world);
```

so far I've tried something like:

```rust
fn hello_world<'mw>(_req: &mut Request, res: Response<'mw, Config>) -> MiddlewareResult<'mw> {
```

but this fails as well with:

```
expected `std::result::Result<nickel::Action<nickel::Response<'mw>, nickel::Response<'mw, (), hyper::net::Streaming>>, nickel::NickelError<'mw>>`,
    found `std::result::Result<nickel::Action<nickel::Response<'_, Config>, nickel::Response<'_, Config, hyper::net::Streaming>>, nickel::NickelError<'_, Config>>`
(expected (),
    found struct `Config`) [E0308]
```

any ideas on how I can use method references with server data?

or any better suggestions on how I can split my app into multiple files - with or without macros? passing a reference to `server` to all my *controller methods* looks a bit ugly so I wanted to keep that as my last option.

I've also tried creating methods that use the `middleware!` macro an box the result. however the `server.get()` method didn't like my `Box`, dereferenced or not.

thanks!

**Comment by lipanski** [comment]
dug a bit deeper in the other issues and [found the answer](https://github.com/nickel-org/nickel.rs/issues/286#issuecomment-155639336) to my question:

```rust
fn hello_world<'mw>(_req: &mut Request<Config>, res: Response<'mw, Config>) -> MiddlewareResult<'mw, Config> {
    res.send("Hello World")
}
```

yay! :dancer:

**Comment by lipanski** [comment]
I've made a PR to add an example for this: https://github.com/nickel-org/nickel.rs/pull/352

**Comment by Ryman** [comment]
Closed via #352. Thanks again for making an example for this!
